### PR TITLE
TD 2069 vehicle select dialog

### PR DESCRIPTION
**Pull request by nikitharani**
Closes [TD-2069](https://sce.myjetbrains.com/youtrack/issue/TD-2069/Create-add-vehicle-dialog)
Figma: https://www.figma.com/proto/B75H0DFTEiUFjJtehIKa8M/VIRTO.MODEL?node-id=522-57017&scaling=scale-down&page-id=0%3A1&starting-point-node-id=228%3A28915

## Changes
- Created a new VehicleSelectDialog component
- Enables the "Add Vehicle" upon valid vehicle selection
- Added callback to return the currently selected vehicles
- Upgraded to node version 18 and upgraded the actions/setup-node action to the latest version
- Moved ActionDialog, DialogTitle, and VehicleSelectDialog components under the Dialog category from General
- Renamed selected vehicles to value, all gates to gates, on vehicle change to onChange, allVehicles to variants to be consistent with vehicleSelect Component
- Added a new prop disableCloseOnSelect to the AutoComplete Component
- Enabled disableCloseOnSelect on multi-select Autocompletes inside VehicleSelect Component
- Because of the disableCloseOnSelect prop which prevents the dropdown from closing when an option is selected (VehicleSlect Component test failed). Modified tests to explicitly close the dropdown after selecting an option using userEvent.click.

## UI/UX
image

Enables the "Add Vehicle" upon valid vehicle selection
image

 The callback returns the currently selected vehicles
image

## Author checklist before assigning a reviewer

- [x] Reviewed my own code-diff.
- [ ] Branch has been run in docker.
- [x] PR assigned to me or an appropriate delegate.
- [x] Relevant labels added to the PR.
- [x] Appropriate tests have been added.
- [x] Lint and test workflows pass.

### CODE REVIEW

**Comment by evoinea** [comment]
@Sowbhagya-ipg can you please have a look on UI/UX changes?

**Comment by Sowbhagya-ipg** [comment]
> @Sowbhagya-ipg can you please have a look on UI/UX changes?

thanks for tagging me. I will do it now

File: src/VehicleSelectDialog/VehicleSelectDialog.test.tsx [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  const saveButton = screen.getByText("Save");
+  fireEvent.click(saveButton);
+
+  // check that the onSaveClick function was called
+  expect(mockOnSaveClick).toHaveBeenCalledWith(selectedVehicles);
+});
+
+test("disables the Save button when any field is empty", () => {
+  // modify the selected vehicles to have an empty gate field
+  const modifiedVehicles = [...selectedVehicles];
```
Comment (evoinea) [inline src/VehicleSelectDialog/VehicleSelectDialog.test.tsx:85]:
```suggestion
  const modifiedVehicles = [...selectedVehicles];
  
```

File: src/VehicleSelectDialog/VehicleSelectDialog.test.tsx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  // empty the gate field
+  modifiedVehicles[0].gate = "";
+
```
Comment (evoinea) [inline src/VehicleSelectDialog/VehicleSelectDialog.test.tsx:88]: Please add missing comment

File: src/VehicleSelectDialog/VehicleSelectDialog.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+const meta: Meta<typeof VehicleSelectDialog> = {
+  component: VehicleSelectDialog,
+  title: "General/VehicleSelectDialog"
+};
+
+export default meta;
+
+const Template: StoryFn<CombinedVehicleProps> = args => {
+  const [open, setOpen] = React.useState(false);
+
```
Comment (evoinea) [inline src/VehicleSelectDialog/VehicleSelectDialog.stories.tsx:23]: Missing comments

File: src/VehicleSelectDialog/VehicleSelectDialog.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+  React.useEffect(() => {
+    updateArgs({ selectedVehicles });
+  }, [selectedVehicles, updateArgs]);
+
+  // callback for when the selected vehicles change
+  const onVehicleChange = (selectedVehicle: SelectedVehicle[]) => {
+    updateArgs({ selectedVehicles: selectedVehicle });
+    action("onVehicleChange")(selectedVehicle);
+  };
+
```
Comment (evoinea) [inline src/VehicleSelectDialog/VehicleSelectDialog.stories.tsx:47]: Missing block comment

**Review by evoinea — CHANGES_REQUESTED** [review]
Few small changes required.

**Comment by evoinea** [comment]
I've deleted the Model Year, but this didn't clear the vehicle variant and the gate; it just disabled them:
image

**Comment by dmbarry86** [comment]
I think the interface for this component has been designed wrong and makes it harded to use than it should be. The dialog itself should be handling the local state when user selects items. You don't need to expose onChange callback from the dialog and manage state outside of the component and then pass in value. You only need onCancelClick and onSaveClick callbacks in this component. Component itself listens to onChange in the VehicleSelect and keeps state in synch then when onSaveClick is called you just need to pass out what the currently selected value is. Otherwise everywhere that uses this component then needs to handle onChange and have external value state. I don't think we ever need to open the dialog with initial values selected so I don't think we need values and onChange props. It's a much simpler interface without them.

**Comment by dmbarry86** [comment]
these props are still showing for me in Storybook

image

**Comment by dmbarry86** [comment]
the props are also still included in the type definition

image

**Review by dmbarry86 — CHANGES_REQUESTED** [review]
- [ ] remove onChange and value from storybook
- [ ] remove onChange and value from types
- [ ] ensure state is reset when dialog is opened

**Review by dmbarry86 — APPROVED** [review]

**Pull request merged**